module/epochs/epoch_lookup_test.go

```diff
@@ -1,65 +1,261 @@
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch, suite.nextEpoch)
+
+	// should handle multiple deliveries of the protocol event
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+	suite.lookup.EpochCommittedPhaseStarted(suite.currentEpochCounter, firstBlockOfCommittedPhase)
+
+	// validate queries are answered correctly
+	testEpochForViewWithFallback(suite.T(), suite.lookup, suite.state, suite.currEpoch, suite.nextEpoch)
+}
+
+// testEpochForViewWithFallback accepts a constructed EpochLookup and state, and
+// validates correctness by issuing various queries, using the input state and
+// epochs as source of truth.
+func testEpochForViewWithFallback(t *testing.T, lookup *EpochLookup, state protocol.State, epochs ...epochRange) {
+	epochFallbackTriggered, err := state.Params().EpochFallbackTriggered()
+	require.NoError(t, err)
+
+	t.Run("should have set epoch fallback triggered correctly", func(t *testing.T) {
+		assert.Equal(t, epochFallbackTriggered, lookup.epochFallbackIsTriggered.Load())
+	})
+
+	t.Run("should be able to query within any committed epoch", func(t *testing.T) {
+		for _, epoch := range epochs {
+			counter, err := lookup.EpochForViewWithFallback(randUint64(epoch.firstView, epoch.finalView))
+			assert.NoError(t, err)
+			assert.Equal(t, epoch.counter, counter)
+		}
+	})
+
+	t.Run("should return ErrViewForUnknownEpoch below earliest epoch", func(t *testing.T) {
+		_, err := lookup.EpochForViewWithFallback(randUint64(0, epochs[0].firstView-1))
+		assert.ErrorIs(t, err, model.ErrViewForUnknownEpoch)
+	})
+
+	// if epoch fallback is triggered, fallback to returning latest epoch counter
+	// otherwise return ErrViewForUnknownEpoch
+	if epochFallbackTriggered {
+		t.Run("should use fallback logic for queries above latest epoch when epoch fallback is triggered", func(t *testing.T) {
+			counter, err := lookup.EpochForViewWithFallback(epochs[len(epochs)-1].finalView + 1)
+			assert.NoError(t, err)
+			// should fallback to returning the counter for the latest epoch
+			assert.Equal(t, epochs[len(epochs)-1].counter, counter)
+		})
+	} else {
+		t.Run("should return ErrViewForUnknownEpoch for queries above latest epoch when epoch fallback is not triggered", func(t *testing.T) {
+			_, err := lookup.EpochForViewWithFallback(epochs[len(epochs)-1].finalView + 1)
+			assert.ErrorIs(t, err, model.ErrViewForUnknownEpoch)
+		})
 	}
 }
+
+// newMockEpoch returns a mock epoch with the given fields set.
+func newMockEpoch(counter, firstView, finalView uint64) *mockprotocol.Epoch {
+	epoch := new(mockprotocol.Epoch)
+	epoch.On("FirstView").Return(firstView, nil)
+	epoch.On("FinalView").Return(finalView, nil)
+	epoch.On("Counter").Return(counter, nil)
+	return epoch
+}
+
+// randUint64 returns a uint64 in [min,max]
+func randUint64(min, max uint64) uint64 {
```

Comment:
useful, lets move this to test utils

Reply:
https://github.com/onflow/flow-go/pull/2927/commits/db064195c49348494d53e642be89f955b7a21ef5